HAL API for CPU UID

MikeDK opened the pull request:
## Description

- Added HAL API for retrieving CPU's unique serial number.
- Added driver class using this new API.
- Added HAL API implementation for STM32, EFM32 and NRF5x targets.
- Added CPU UID ability into targets.json for STM32, EFM32 and NRF5x targets

## Status

**READY**

## Todos

Other chip vendors than STM, Silicon Labs or Nordic should also provide an implementation of this HAL.

I am also not sure about the licenses - In the target implementation files I used the License headers of the corresponding vendors, but as I do not work for those companies, I am not sure if those lines fit ;)

## Related Issue

#5544

MikeDK committed:
Added HAL API for retrieving the CPU UID

MikeDK committed:
Added driver class using the CPU UID HAL API

MikeDK committed:
Added CPU UID target implementation for STM32 devices

MikeDK committed:
Added CPU UID target implementation for EFM32 devices

MikeDK committed:
Added CPU UID target implementation for NRF5x devices

MikeDK committed:
Added CPU UID ability to STM32, EFM32 and NRF5x targets

bulislaw reviewed (commented):
Did you have any thoughts about how to test this?

MikeDK committed:
CpuUid: to save memory, overloaded type cast std::string operator uses minimal implementation instead of snprintf

MikeDK committed:
CpuUid: buffer size assertion added

Added note in hal that a value higher than 0 is expected from the target implementation for the cpu_uid_get_length() function.

Added MBED_ASSERT for checking cpu_uid_get_length() return value in the driver.

bulislaw reviewed (approved):
LGTM
CC: @bremoran @bogdanm @LiyouZhou @seankinghan

MikeDK committed:
CpuUid: fixed coding style

MikeDK committed:
CpuUid: fixed missing include for MBED_ASSERT()

MikeDK committed:
CpuUid: moved includes of string and vector

The problem appeared in the travis-ci build when building the SAMR21G18A target. It seems this target lacks support of STL somehow.
If DEVICE_CPUUID is not defined, the string and vector headers would still be included - so on "buggy stl" targets, this would lead to build error

bcostm reviewed (approved):
LGTM

sg- reviewed (changes requested):
Looking good! This should be target to a feature branch for partner implementations and not come straight to master. one of the @bulislaw should be able to create one for you.

MikeDK committed:
CpuUid: complete refactoring to meet review requirements

Hopefully all review items were addressed.

MikeDK committed:
CpuUid: removed STL includes, removed desctructor definition, as default destructor will be enough, and added critical sections to code filling buffers

MikeDK committed:
CpuUid: build error for FAMILY_STM32 derived targets

some targets derived from FAMILY_STM32 define a "macros" part in their target definition instead of "macros_add", so the macro definition in FAMILY_STM32 gets overwritten.
Changed to "macros_add" for these devices.

MikeDK committed:
CpuUid: fixed build error in target NRF51822

sg- reviewed (dismissed):
Looking quite a bit better. Still dont like the macros MBED_CPU_UID_SIZE and  MBED_CPU_UID_STR_SIZE_MAX.
Why cant the HAL return the size of this from the device and the driver cope with the size if it needs a local buffer? In fact, why have so many buffers? Each accessor seems to call to the hardware and has a buffer just to pass back to the caller. Could the caller not provide the buffers needed? I dont fully understand how this is used further into the application wrt storage needs.

SenRamakri reviewed (dismissed):
Please look into my comments.

MikeDK committed:
CpuUid: renamed files to device uid naming scheme

MikeDK committed:
CpuUid: renamed all references in code to device uid

MikeDK committed:
DeviceUid: changed uid vendor string in STM targets to a much smaller form

LMESTM reviewed (approved):
When it comes to STM32 targets, this looks good to me !

0xc0170 reviewed (approved):
I had previously 2 questions, that should not block this

LGTM

MikeDK committed:
changed warnings in device_uid_api.h to be thrown also when size or str_size_max are zero

MikeDK committed:
fixed missing include in CriticalSectionLock.h (macro MBED_DEPRECATED_SINCE not defined)

jeromecoutant commented:
Hi
Still no date for this feature branch merge...?

cmonr commented:
@bulislaw Would you happen to know?

bulislaw commented:
I'm thinking having all STM, SiLabs and Nordic boards implemented is a good start. I would say we should try to make a push and merge it in 5.13, otherwise it only sits here and gets out of date. Using TAM team to possibly get other vendors to contribute.

@c1728p9 @donatieng Any thoughts?

bulislaw commented:
> I'm thinking having all STM, SiLabs and Nordic boards implemented is a good start. I would say we should try to make a push and merge it in 5.13, otherwise it only sits here and gets out of date. Using TAM team to possibly get other vendors to contribute.

@c1728p9 @donatieng Any thoughts?

c1728p9 commented:
Seems reasonable to me

bulislaw commented:
@c1728p9 could you create a PREQ for that in 5.13 please.

c1728p9 commented:
I created it and added you as a watcher.
